google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -736,29 +749,51 @@ BatchState getState() {
     return state;
   }
 
-  boolean hasDocument(DocumentReference documentReference) {
-    return pendingOperations.containsKey(documentReference);
+  List<String> getPendingDocumentPaths() {
+    return FluentIterable.from(pendingOperations)
+        .transform(
+            new Function<PendingOperation, String>() {
+              @Override
+              public String apply(PendingOperation input) {
+                return input.key;
+              }
+            })
+        .toList();
   }
 
-  Set<DocumentReference> getPendingDocuments() {
-    return pendingOperations.keySet();
+  List<Integer> getPendingIndexes() {
+    return FluentIterable.from(pendingOperations)
+        .transform(
+            new Function<PendingOperation, Integer>() {
+              @Override
+              public Integer apply(PendingOperation input) {
+                return input.writeBatchIndex;
+              }
+            })
+        .toList();
   }
 
-  Collection<SettableApiFuture<WriteResult>> getPendingFutures() {
-    return pendingOperations.values();
+  List<SettableApiFuture<WriteResult>> getPendingFutures() {
+    return FluentIterable.from(pendingOperations)
+        .transform(
+            new Function<PendingOperation, SettableApiFuture<WriteResult>>() {
+              @Override
+              public SettableApiFuture<WriteResult> apply(PendingOperation input) {
+                return input.future;
+              }
+            })
+        .toList();
   }
 
   int getPendingOperationCount() {
     return pendingOperations.size();
   }
 
   private ApiFuture<WriteResult> processOperation(DocumentReference documentReference) {
-    Preconditions.checkState(
-        allowDuplicateDocs() || !pendingOperations.containsKey(documentReference),
-        "Batch should not contain writes to the same document");
     Preconditions.checkState(state == BatchState.OPEN, "Batch should be OPEN when adding writes");
     SettableApiFuture<WriteResult> result = SettableApiFuture.create();
-    pendingOperations.put(documentReference, result);
+    pendingOperations.add(
+        new PendingOperation(getPendingOperationCount(), documentReference.getPath(), result));
```

Comment:
```suggestion
    int nextBatchIndex = getPendingOperationCount();
    pendingOperations.add(
        new PendingOperation(nextBatchIndex, documentReference.getPath(), result));
```

Optional, but it does remove the need to understand the execution order.

Reply:
done.